Mathlib/Tactic/HaveLetLinter.lean

```diff
@@ -0,0 +1,99 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import Lean.Elab.Command
+import Lean.Linter.Util
+import Std.Data.List.Basic
+
+/-!
+#  The `have` vs `let` linter
+
+The `have` vs `let` linter flags uses of `have` to introduce a hypothesis whose Type is not `Prop`.
+
+TODO:
+* `replace` implies `have`: should it be ignored?
+* `haveI` may need to change to `let/letI`?
+* also do `let` vs `have`.
+-/
+
+open Lean Elab Command Meta
+
+namespace Mathlib.Linter
+
+/-- The `have` vs `let` linter emits a warning on `have`s introducing a hypothesis whose
+Type is not `Prop`. -/
+register_option linter.haveLet : Bool := {
+  defValue := true
+  descr := "enable the `have` vs `let` linter"
+}
+
+namespace haveLet
+
+/-- find the `have` syntax. -/
+partial
+def isHave? : Syntax → Bool
+  | .node _ ``Lean.Parser.Tactic.tacticHave_ _ => true
+  |_ => false
+
+/-- `SyntaxNodeKind`s that imply a `have` but should be ignored anyway. -/
+abbrev exclusions : HashSet SyntaxNodeKind := HashSet.empty
+  |>.insert ``Lean.Parser.Tactic.replace
+  |>.insert `Std.Tactic.classical!
+  |>.insert `Mathlib.Tactic.Tauto.tauto
+
+end haveLet
+
+end Mathlib.Linter
+
+namespace Mathlib.Linter.haveLet
+
+/-- returns the `have` syntax whose corresponding hypothesis does not have Type `Prop` and
+also a `Format`ted version of the corresponding Type. -/
+partial
+def nonPropHaves : InfoTree → CommandElabM (Array (Syntax × Format))
+  | .node i args => do
+    let nargs := (← args.toArray.mapM nonPropHaves).flatten
+    if let .ofTacticInfo i := i then
+      let stx := i.stx
+      if exclusions.contains stx.getKind then return #[] else
+      if isHave? stx then
+        let mctx := i.mctxAfter
+        let mvdecls := (i.goalsAfter.map (mctx.decls.find? ·)).reduceOption
+        let _ : Ord MetavarDecl := { compare := (compare ·.index ·.index) }
+        -- we extract the `MetavarDecl` with largest index after a `have`, since this one
+        -- holds information about the metavariable where `have` introduces the new hypothesis.
+        let largestIdx := mvdecls.toArray.qsort (·.index > ·.index)
+        -- the relevant `LocalContext`
+        let lc := (largestIdx.getD 0 default).lctx
+        -- and the last declaration introduced: the one that `have` created
+        let ld := (lc.lastDecl.getD default).type
+        -- now, we get the `MetaM` state up and running to find the type of `ld`
+        let res ← liftCoreM do MetaM.run (ctx := { lctx := lc }) (s := { mctx := mctx }) <| do
```

Comment:
As discussed in the other PR; I think you want `ContextInfo.runMetaM` here. You can get the context you need by using [Lean.Elab.InfoTree.foldInfoTree](https://leanprover-community.github.io/mathlib4_docs/Lean/Server/InfoUtils.html#Lean.Elab.InfoTree.foldInfoTree)

Reply:
Eric, I tried to use this, but I cannot see how.  From the `InfoTree`s, I could get a `PartialContextInfo`, produce a `ContextInfo` using `.mergeIntoOuter?`, but then to use `Lean.Elab.InfoTree.foldInfo` I still need a `LocalContext` (that I could probably fetch inside the `MetavarContext` embedded into to `ContextInfo`), but I do not know how to find which goals are post-`have` and which ones are pre-`have`, since they seem to be all mashed together.

On the other hand, if I am at a `.node` of the InfoTree`, I can do what I did, but there I do not seem to have access to a `(Partial)ContextInfo`, so I would not know where to start...

Reply:
I came up with this:
```lean
def InfoTree.foldInfoM {α m} [Monad m] (f : ContextInfo → Info → α → m α) (init : α) :
    InfoTree → m α :=
  InfoTree.foldInfo (fun ctx i ma => do f ctx i (← ma)) (pure init)

/-- returns the `have` syntax whose corresponding hypothesis does not have Type `Prop` and
also a `Format`ted version of the corresponding Type. -/
partial
def nonPropHaves : InfoTree → CommandElabM (Array (Syntax × Format)) :=
  InfoTree.foldInfoM
    (fun ctx info args => return args ++ (← do
      let .ofTacticInfo i := info | return #[]
      let stx := i.stx
      if exclusions.contains stx.getKind then return #[]
      unless isHave? stx do return #[]
      let mctx := i.mctxAfter
      let mvdecls := (i.goalsAfter.map (mctx.decls.find? ·)).reduceOption
      let _ : Ord MetavarDecl := { compare := (compare ·.index ·.index) }
      -- we extract the `MetavarDecl` with largest index after a `have`, since this one
      -- holds information about the metavariable where `have` introduces the new hypothesis.
      let largestIdx := mvdecls.toArray.qsort (·.index > ·.index)
      -- the relevant `LocalContext`
      let lc := (largestIdx.getD 0 default).lctx
      -- and the last declaration introduced: the one that `have` created
      let ld := (lc.lastDecl.getD default).type
      -- now, we get the `MetaM` state up and running to find the type of `ld`

      ctx.runMetaM lc do
        let typ ← inferType (← instantiateMVars ld)
        if ! typ.isProp then
          return #[(stx, ← ppExpr ld)]
        else
          return #[]))
    #[]
```

Reply:
Eric, I factored out your monadic `foldInfo`: I am still trying to wrap my head around it, but it looks great, thanks!